tool-openssl/dgst.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <fcntl.h>
+#include <iostream>
+
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include "internal.h"
+
+// MD5 command currently only supports stdin
+static const argument_t kArguments[] = {
+    {"-help", kBooleanArgument, "Display option summary"},
+    {"-hmac", kOptionalArgument,
+     "Create a hashed MAC with the corresponding key"},
+    {"", kOptionalArgument, ""}};
+
+static bool dgst_file_op(const std::string &filename, const EVP_MD *digest,
+                         const std::string &hmac_key) {
+  ScopedFD scoped_fd = OpenFD(filename.c_str(), O_RDONLY | O_BINARY);
+  int fd = scoped_fd.get();
+
+  static const size_t kBufSize = 8192;
+  std::unique_ptr<uint8_t[]> buf(new uint8_t[kBufSize]);
+
+  if (!hmac_key.empty()) {
+    bssl::ScopedHMAC_CTX ctx;
+    if (!HMAC_Init_ex(ctx.get(), hmac_key.data(), hmac_key.size(), digest,
+                      nullptr)) {
+      fprintf(stderr, "Failed to initialize HMAC_Init_ex.\n");
+      return false;
+    }
+
+    // Update |buf| from file continuously.
+    for (;;) {
+      size_t n;
+      if (!ReadFromFD(fd, &n, buf.get(), kBufSize)) {
+        fprintf(stderr, "%s: No such file or directory\n", filename.c_str());
+        return false;
+      }
+
+      if (n == 0) {
+        break;
+      }
+
+      if (!HMAC_Update(ctx.get(), buf.get(), n)) {
+        fprintf(stderr, "Failed to update HMAC.\n");
+        return false;
+      }
+    }
+
+    const unsigned expected_mac_len = EVP_MD_size(digest);
+    std::unique_ptr<uint8_t[]> mac(new uint8_t[expected_mac_len]);
+    unsigned mac_len;
+    if (!HMAC_Final(ctx.get(), mac.get(), &mac_len)) {
+      fprintf(stderr, "Failed to finalize HMAC.\n");
+      return false;
+    }
+
+    // Print HMAC output.
+    fprintf(stdout, "HMAC-%s(%s)= ", EVP_MD_get0_name(digest),
```

Comment:
NP: This would display the full path.  Is that expected?

Reply:
Yes that's expected. OpenSSL's output example:
```
aws-lc on  main [?] via △ v3.20.0 via  v1.21.3 
❯ openssl dgst $PWD/../bio_test
SHA256(/Users/sachiang/Documents/repo/aws-lc/../bio_test)= 28176a1f8de724adeb381febacf90595715b55a15294c14c24a3e5f96ced04dc
```